# Study Case

Dilakukan data preparation dengan menggunakan data ride_sharing_new.csv, yaitu data mengenai penggunaan layanan peminjaman sepeda oleh user 

# GDrive Connection

In [25]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [26]:
# Load the data
path ="/content/drive/MyDrive/Data Analyst/Meet 13/Data/ride_sharing_new.csv"

# Import Library

In [27]:
import pandas as pd
import os

# Case 1

Informasi-informasi terkait data ride_sharing_new.

In [8]:
df =pd.read_csv(path)

In [9]:
df.head()

duration  station_A_id  \
0  12 minutes            81   
1  24 minutes             3   
2   8 minutes            67   
3   4 minutes            16   
4  11 minutes            22   

                                      station_A_name  station_B_id  \
0                                 Berry St at 4th St           323   
1       Powell St BART Station (Market St at 4th St)           118   
2  San Francisco Caltrain Station 2  (Townsend St...            23   
3                            Steuart St at Market St            28   
4                              Howard St at Beale St           350   

                    station_B_name  bike_id  user_type  user_birth_year  \
0               Broadway at Kearny   5480.0          2             1959   
1  Eureka Valley Recreation Center   5193.0          2             1965   
2    The Embarcadero at Steuart St   3652.0          3             1993   
3     The Embarcadero at Bryant St   1883.0          1             1979   
4             8th St at Brannan St   4626.0          2             1994   

  user_gender  
0        Male  
1        Male  
2        Male  
3        Male  
4        Male

Data ride_sharing_new terdiri dari beberapa kolom antara lain mengenai durasi pemakaian sepeda oleh user, nama stasiun keberangkatan (station_A_name) dan nama stasiun pemberhentian (station_B_name), kemudian informasi mengenai gender user dan tahun kelahiran.

In [10]:
#cek tipe data, dan jumlah data secara keseluruhan 
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25760 entries, 0 to 25759
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   duration         25760 non-null  object 
 1   station_A_id     25760 non-null  int64  
 2   station_A_name   25760 non-null  object 
 3   station_B_id     25760 non-null  int64  
 4   station_B_name   25760 non-null  object 
 5   bike_id          25755 non-null  float64
 6   user_type        25760 non-null  int64  
 7   user_birth_year  25760 non-null  int64  
 8   user_gender      25760 non-null  object 
dtypes: float64(1), int64(4), object(4)
memory usage: 1.8+ MB


Dilihat dari info data di atas, terdapat perbedaan jumlah data pada kolom bike_id. Hal ini bisa mengindikasikan bahwa mungkin saja terdapat missing value pada kolom tersebut 

Kemudian agaimana Summary Statistics dari data tersebut ?



In [11]:
df.describe()

station_A_id  station_B_id       bike_id     user_type  user_birth_year
count  25760.000000  25760.000000  25755.000000  25760.000000     25760.000000
mean      31.023602     89.558579   4108.053388      2.008385      1983.054969
std       26.409263    105.144103   1576.015656      0.704541        10.010992
min        3.000000      3.000000     11.000000      1.000000      1901.000000
25%       15.000000     21.000000   3108.000000      2.000000      1978.000000
50%       21.000000     58.000000   4821.000000      2.000000      1985.000000
75%       67.000000     93.000000   5258.000000      3.000000      1990.000000
max       81.000000    383.000000   6638.000000      3.000000      2001.000000

# Case 2

Cek apakah terdapat Missing Values ?

In [12]:
jumlah_missing = pd.isnull(df).sum()
jumlah_missing

duration           0
station_A_id       0
station_A_name     0
station_B_id       0
station_B_name     0
bike_id            5
user_type          0
user_birth_year    0
user_gender        0
dtype: int64

Terdapat missing value sebanyak 5 data pada kolom bike_id. Namun data tersebut termasuk dalam variabel yang tidak bisa digantikan dengan imputasi data karena setiap peminjaman sepeda harus tercatat ID sepeda yang digunakan, missing values pada bike_id mungkin terjadi karena adanya kesalahan dalam pencatatan. Terdapat 2 alternatif dalam menyelesaikan kasus ini yaitu mencari tau ID sepeda yang seharusnya dicatat atau tidak memasukkan data yang tidak memiliki ID sepeda dalam analisis kali ini.

In [13]:
# melakukan DROP pada data missing values
df.dropna(axis=0,how='any',inplace=True)

In [14]:
# cek missing value kembali
df.isnull().sum()

duration           0
station_A_id       0
station_A_name     0
station_B_id       0
station_B_name     0
bike_id            0
user_type          0
user_birth_year    0
user_gender        0
dtype: int64

In [16]:
df.shape

(25755, 9)

Setelah kita hapus maka jumlah data yang dipunya sekarang adalah 25755 baris dan 9 kolom 

# Case 3

Cek apakah terdapat data duplicate? 

In [17]:
df[df.duplicated(keep =False)].sort_values("bike_id")

duration  station_A_id  \
9695   17 minutes            22   
25757  17 minutes            22   
566     9 minutes            81   
25756   9 minutes            81   
7609   10 minutes            30   
25758  10 minutes            30   
9723    4 minutes            21   
25759   4 minutes            21   

                                         station_A_name  station_B_id  \
9695                              Howard St at Beale St           102   
25757                             Howard St at Beale St           102   
566                                  Berry St at 4th St            81   
25756                                Berry St at 4th St            81   
7609     San Francisco Caltrain (Townsend St at 4th St)             6   
25758    San Francisco Caltrain (Townsend St at 4th St)             6   
9723   Montgomery St BART Station (Market St at 2nd St)           343   
25759  Montgomery St BART Station (Market St at 2nd St)           343   

                      station_B_name  bike_id  user_type  user_birth_year  \
9695              Irwin St at 8th St    492.0          3             1961   
25757             Irwin St at 8th St    492.0          3             1961   
566               Berry St at 4th St   1225.0          2             1993   
25756             Berry St at 4th St   1225.0          2             1993   
7609   The Embarcadero at Sansome St   4442.0          1             1967   
25758  The Embarcadero at Sansome St   4442.0          1             1967   
9723             Bryant St at 2nd St   5034.0          2             1993   
25759            Bryant St at 2nd St   5034.0          2             1993   

      user_gender  
9695       Female  
25757      Female  
566          Male  
25756        Male  
7609         Male  
25758        Male  
9723         Male  
25759        Male

Ternyata terdapat 4 data duplicate, maka selanjutnya kita akan menghilangkan duplicate tersebut 

In [18]:
df = df.drop_duplicates().reset_index(drop=True)

In [19]:
#Cek kembali apakah masih ada data duplicate 
df[df.duplicated(keep =False)].sort_values("bike_id")

Empty DataFrame
Columns: [duration, station_A_id, station_A_name, station_B_id, station_B_name, bike_id, user_type, user_birth_year, user_gender]
Index: []

In [20]:
df.shape

(25751, 9)

Setelah 4 data duplicate dihilangkan, maka data yang dimiliki sekarang berjumlah 25751 baris dan 9 kolom 

# Case 4

Data pada kolom duration terdapat satuan waktu yang dapat menghalangi kita dalam menganalisis durasi pengendara sepeda. Oleh karena itu, dibutuhkan cleaning data satuan waktu.

In [28]:
df['duration'] = df['duration'].replace({' minutes':''}, regex= True)

In [29]:
df["duration"]

0        12  
1        24  
2         8  
3         4  
4        11  
         ... 
25746    11  
25747    10  
25748    14  
25749    14  
25750    29  
Name: duration, Length: 25751, dtype: object

# Case 5

Mengubah tipe data ke tipe data yang lebih sesuai untuk memudahkan dalam analisa nantinya 

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25751 entries, 0 to 25750
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   duration         25751 non-null  object 
 1   station_A_id     25751 non-null  int64  
 2   station_A_name   25751 non-null  object 
 3   station_B_id     25751 non-null  int64  
 4   station_B_name   25751 non-null  object 
 5   bike_id          25751 non-null  float64
 6   user_type        25751 non-null  int64  
 7   user_birth_year  25751 non-null  int64  
 8   user_gender      25751 non-null  object 
dtypes: float64(1), int64(4), object(4)
memory usage: 1.8+ MB


Disini tipe data untuk duration akan diubah menjadi integer, sedangakan user_type, station_A_id, station_B_id, dan bike_id akan diubah menjadi object/string 

In [54]:
df["duration"] = df["duration"].astype(int)
df["user_type"]= df["user_type"].astype(str)
df["station_A_id"]= df["station_A_id"].astype(str)
df["station_B_id"]= df["station_B_id"].astype(str)
df["bike_id"]= df["bike_id"].astype(str)

In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25751 entries, 0 to 25750
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   duration         25751 non-null  int64 
 1   station_A_id     25751 non-null  object
 2   station_A_name   25751 non-null  object
 3   station_B_id     25751 non-null  object
 4   station_B_name   25751 non-null  object
 5   bike_id          25751 non-null  object
 6   user_type        25751 non-null  object
 7   user_birth_year  25751 non-null  int64 
 8   user_gender      25751 non-null  object
 9   user_age         25751 non-null  int64 
 10  age_class        25751 non-null  object
dtypes: int64(3), object(8)
memory usage: 2.2+ MB


# Case 6

Menghitung jumlah pengendara sepeda yang melewati masing-masing station A dan station B menggunakan groupby()

In [36]:
df_group1= df[['station_A_name','station_B_name','bike_id']].groupby(['station_A_name','station_B_name']).count()
df_group1

bike_id
station_A_name          station_B_name                                             
Berry St at 4th St      11th St at Bryant St                                      3
                        11th St at Natoma St                                     42
                        14th St at Mission St                                     4
                        15th St at Potrero Ave                                   12
                        16th St Mission BART                                      3
...                                                                             ...
Steuart St at Market St Washington St at Kearny St                               45
                        Webster St at Grove St                                    8
                        Webster St at O'Farrell St                                9
                        Williams Ave at 3rd St                                    1
                        Yerba Buena Center for the Arts (Howard St at 3...       16

[1235 rows x 1 columns]

Menghitung jumlah pengendara sepeda berdasarkan user_type dan user_gender.

In [37]:
df_group2 = df[['user_type','user_gender','bike_id']].groupby(['user_type','user_gender']).count()
df_group2

bike_id
user_type user_gender         
1         Female          1465
          Male            4734
          Other             83
2         Female          3044
          Male            9746
          Other            178
3         Female          1516
          Male            4895
          Other             90

# Case 7

Menghitung rata-rata durasi dalam pemakaian sepeda

In [41]:
rata_durasi= df['duration'].mean()
print(f"rata-rata pemakaian sepeda yaitu {rata_durasi} minutes")

rata-rata pemakaian sepeda yaitu 11.388761601491204 minutes


Menghitung rata-rata durasi pemakaian sepeda pada masing-masing rute

In [ ]:
df[['station_A_name','station_B_name','duration']].groupby(['station_A_name','station_B_name']).mean()

duration
station_A_name          station_B_name                                               
Berry St at 4th St      11th St at Bryant St                                11.333333
                        11th St at Natoma St                                10.857143
                        14th St at Mission St                               16.500000
                        15th St at Potrero Ave                               8.083333
                        16th St Mission BART                                15.333333
...                                                                               ...
Steuart St at Market St Washington St at Kearny St                           7.133333
                        Webster St at Grove St                              16.625000
                        Webster St at O'Farrell St                          17.222222
                        Williams Ave at 3rd St                              31.000000
                        Yerba Buena Center for the Arts (Howard St at 3...   9.875000

[1235 rows x 1 columns]

# CASE 8


Pemilik sepeda ingin memberikan promo kepada customer/user yang bertujuan untuk meningkatkan ketertarikan pemuda maupun orang tua dalam bersepeda. Oleh karena itu pemilik sepeda ingin mengetahui rata-rata usia dan modus usia customer sehingga promo yang diberikan tepat.

In [42]:
from datetime import datetime as dt

In [44]:
#mengkonversi informasi birthyear_menjadi umur 
df["user_age"] = dt.now().year - df["user_birth_year"]
df.head()

duration station_A_id                                     station_A_name  \
0        12           81                                 Berry St at 4th St   
1        24            3       Powell St BART Station (Market St at 4th St)   
2         8           67  San Francisco Caltrain Station 2  (Townsend St...   
3         4           16                            Steuart St at Market St   
4        11           22                              Howard St at Beale St   

  station_B_id                   station_B_name bike_id user_type  \
0          323               Broadway at Kearny  5480.0         2   
1          118  Eureka Valley Recreation Center  5193.0         2   
2           23    The Embarcadero at Steuart St  3652.0         3   
3           28     The Embarcadero at Bryant St  1883.0         1   
4          350             8th St at Brannan St  4626.0         2   

   user_birth_year user_gender  user_age  
0             1959        Male        63  
1             1965        Male        57  
2             1993        Male        29  
3             1979        Male        43  
4             1994        Male        28

In [46]:
#Mencari rata rata umur customer 
rata_umur = df["user_age"].mean()
print(f"rata-rata umur pemakai sepeda adalah {round(rata_umur)} tahun")

rata-rata umur pemakai sepeda adalah 39 tahun


In [48]:
df["user_age"].max()

121

In [49]:
df["user_age"].min()

21

In [50]:
#mengelompokan usia
def age_class (age):
  if age >= 65:
    return "lanjut usia"
  elif age >=45:
    return "dewasa"
  elif age >= 12:
    return "remaja"
  else :
    return "anak anak"

In [51]:
df ["age_class"] = df["user_age"].apply(age_class)

In [52]:
df.head()

duration station_A_id                                     station_A_name  \
0        12           81                                 Berry St at 4th St   
1        24            3       Powell St BART Station (Market St at 4th St)   
2         8           67  San Francisco Caltrain Station 2  (Townsend St...   
3         4           16                            Steuart St at Market St   
4        11           22                              Howard St at Beale St   

  station_B_id                   station_B_name bike_id user_type  \
0          323               Broadway at Kearny  5480.0         2   
1          118  Eureka Valley Recreation Center  5193.0         2   
2           23    The Embarcadero at Steuart St  3652.0         3   
3           28     The Embarcadero at Bryant St  1883.0         1   
4          350             8th St at Brannan St  4626.0         2   

   user_birth_year user_gender  user_age age_class  
0             1959        Male        63    dewasa  
1             1965        Male        57    dewasa  
2             1993        Male        29    remaja  
3             1979        Male        43    remaja  
4             1994        Male        28    remaja

In [56]:
#jumlah customer tiap kategori usia
df[["age_class","bike_id"]].groupby(["age_class"]).count()

bike_id
age_class           
dewasa          5718
lanjut usia      435
remaja         19598

Dari pengelompokan tersebut didapatkan bahwa pemakai layanan peminjaman sepeda mayoritas adalah customer dengan kategori umur remaja yang berusia antara 12-45 tahun 